### PR TITLE
WP-r48390: Upgrade/install: Allow plugin and theme updates from a upl…

### PR DESCRIPTION
**Pull request by mattyrob**
…oaded .zip file.

WP:Props mariovalney, cyberhobo, imath, shaunandrews, mariovalney, earnjam, desrosj, dd32, folletto, swissspidy, melchoyce, pento, joshuawold, psykro, clorith, ahortin, galbaras, pingram3541, joyously, doobeedoo, karmatosed, poena, whyisjake, earnjam, sergeybiryukov, audrasjb, azaozz.

Fixes https://core.trac.wordpress.org/ticket/9757.

Conflicts:
- src/wp-admin/css/themes.css
- src/wp-admin/includes/class-plugin-installer-skin.php
- src/wp-admin/includes/class-plugin-upgrader.php
- src/wp-admin/includes/class-theme-installer-skin.php
- src/wp-admin/includes/class-theme-upgrader.php
- src/wp-admin/includes/class-wp-upgrader.php
- src/wp-admin/update.php

---

Merges https://core.trac.wordpress.org/changeset/48390 / WordPress/wordpress-develop@289c28637c to ClassicPress.

## Description
This backport of upstream patch allows plugins and themes to be updated from manually uploaded zip files

## Motivation and context
This is a long time requested enhancement in the WordPress core.
I have attempted to resolve all conflicting areas within the patch

## How has this been tested?
Upstream commit
## Types of changes
- New feature

### CODE REVIEW

**Comment by bahiirwa** [comment]
This is really useful. Top on my wishlist.

**Comment by bahiirwa** [comment]
This Backport falls in line with this [Petition](https://petitions.classicpress.net/posts/132/update-themes-or-plugins-through-the-installer). Could this be considered for the next version release?

**Comment by nylen** [comment]
Yes, I plan to include this in v1.2.0.

Here's what needs to be done first:

- Create one commit with the merge conflicts and then resolve the conflicts in a separate commit, as the latest version of the backport script does. This allows reviewing the conflict resolution too which is an important step.
- Testing

**Comment by mattyrob** [comment]
@Nylen I cannot recall what the conflicts were now but I think they were all syntax / layout based.

Is it better to close and delete this PR and start over or can it be rolled back?

**Comment by nylen** [comment]
Much better to re-use the same PR, otherwise the activity related to this change gets hard to follow. There's also no need to re-do your conflict resolution, having that already done is very helpful. I will insert the first commit with the conflict markers when I have a chance.

**Comment by nylen** [comment]
This is going to need a good bit more work and it's going to pull in a bunch of other stuff. There are 10 commits linked to https://core.trac.wordpress.org/ticket/9757 and at least some of them require functions that we haven't introduced yet:

`git log wp/master -Sis_wp_version_compatible --oneline`

- WordPress/wordpress-develop@9636a9d4d7 Upgrade/Install: Check WordPress and PHP version requirements when uploading a plugin or theme from ZIP file.
- WordPress/wordpress-develop@deb78bdf1c Themes: Display a message on Themes list table if a theme update requires a higher version of PHP or WordPress.
- WordPress/wordpress-develop@6f79ff4d1c Themes: Display a message in theme updates list if an update requires a higher version of PHP or WordPress.
- WordPress/wordpress-develop@7ea44b5add Themes: Display a message in theme grid and Theme Details modal if a theme update requires a higher version of PHP or WordPress.
- WordPress/wordpress-develop@51e92ced21 Upgrade/Install: Use `is_php_version_compatible()` and `is_wp_version_compatible()` in plugin and theme requirement checks.
- WordPress/wordpress-develop@a60b30444f Themes: Prevent installation of themes that require a higher version of PHP or WordPress.
- WordPress/wordpress-develop@b91e74d2f6 Themes: Prevent activation and live preview of themes that require a higher version of PHP or WordPress.
- WordPress/wordpress-develop@286d672389 Themes: Block theme activation if it requires a higher version of PHP or WordPress.
- WordPress/wordpress-develop@78e868e4c7 Plugins: Introduce `is_wp_version_compatible()` and `is_php_version_compatible()` for checking compatibility with the current WordPress or PHP version.

Note commits like "Prevent installation of plugins/themes that require a higher version of PHP or WordPress". This is starting to become a problem for existing ClassicPress installations so we might want to consider enabling that behavior also.

**Comment by nylen** [comment]
Due to the additional number of changesets to be backported from WP, and the connections of some of these changesets to the new behavior we also want to adopt (Prevent installation of plugins/themes that require a higher version of PHP or WordPress), this isn't going to be able to make it into version `1.2.0`.

**Comment by mattyrob** [comment]
I seem to have introduced a syntax error and broken things.

I think given the number of tickets and changes it may be prudent to apply the changes in order and open a new branch - if that's okay I'll aim to do that over the next week or so.

**Comment by mattyrob** [comment]
@nylen 

Please see update branch and Pull Request:
https://github.com/ClassicPress/ClassicPress/pull/621

I have tested it locally but it needs more testing and a review.

I'm leaving this open for now but it can be closed if the new pull request is better to take forward.

**Comment by mattyrob** [comment]
Please see pull #621 

Closing this PR in favour of the better PR
